core/include/cubos/core/gl/render_device.hpp

```diff
@@ -1454,6 +1479,94 @@ namespace cubos::core::gl
         protected:
             ShaderBindingPoint() = default;
         };
+
+        /// @brief Abstract timer.
+        class CUBOS_CORE_API Timer
+        {
+        public:
+            virtual ~Timer() = default;
+
+            /// @brief Starts a region to be timed.
+            virtual void begin() = 0;
+
+            /// @brief Ends a region.
+            virtual void end() = 0;
+
+            /// @brief Returns whether the result is available.
+            /// @return True if the result is available, false otherwise.
+            virtual bool done() = 0;
+
+            /// @brief Gets the result, in nanoseconds.
+            ///
+            /// Blocks until a result is available.
+            ///
+            /// @return Time spent in region, in nanoseconds.
+            virtual int result() = 0;
+
+        protected:
+            Timer() = default;
+        };
+
+        /// @brief Pipelined timer.
+        class CUBOS_CORE_API PipelinedTimer
+        {
+        public:
+            struct Result
+            {
+                int result;          ///< Time spent in region in the latest frame that finished rendering.
+                size_t framesBehind; ///< How many frames behind the CPU the latest rendered frame lags.
+            };
+
+            PipelinedTimer(RenderDevice* rd)
+                : mRd(rd)
+            {
+            }
+
+            ~PipelinedTimer() = default;
+
+            /// @brief Starts a region to be timed.
+            void begin()
+            {
+                if (mNeedsEnd) // make sure it doesn't begin again before ending
+                {
+                    return;
+                }
```

Comment:
I would `CUBOS_ASSERT` this, as when this happens it's always a mistake on the programmer's side right?
```suggestion
                CUBOS_ASSERT (!mNeedsEnd, "The previous call to begin() wasn't matched with a call to end()");
```